### ISSUE TITLE
[samsungtv] TV always offline, although detected initially

### ISSUE THREAD

**Issue by martin5233** [openhab/openhab-addons#17511]
I have a Samsung UE32ES6300, which used to work with the SamsungTV binding in OpenHAB 3. Since I upgraded to OpenHAB 4 about a year ago I could never get these two to work together again.

## Expected Behavior
The binding should detect the TV and change the thing's state to online, when turning on the TV

## Current Behavior
The binding does detect the TV, when the thing is first added. The following lines are logged:
```
2024-10-05 12:13:24.138 [DEBUG] [.samsungtv.internal.WakeOnLanUtility] - 192.168.1.35: os: linux
2024-10-05 12:13:24.263 [DEBUG] [.samsungtv.internal.WakeOnLanUtility] - 192.168.1.35: MAC address of host 192.168.1.35 is dc:71:44:52:cc:eb
2024-10-05 12:13:24.266 [DEBUG] [tv.internal.handler.SamsungTvHandler] - 192.168.1.35: Updated Configuration macAddress:dc:71:44:52:cc:eb
2024-10-05 12:13:24.268 [TRACE] [tv.internal.handler.SamsungTvHandler] - 192.168.1.35: getting TV properties
2024-10-05 12:13:24.274 [WARN ] [tv.internal.handler.SamsungTvHandler] - 192.168.1.35: Cannot connect to TV: java.util.concurrent.ExecutionException: java.net.ConnectException: Connection re
fused                                                                                                                                                                                         
2024-10-05 12:13:24.274 [WARN ] [tv.internal.handler.SamsungTvHandler] - 192.168.1.35: Cannot get TVProperties, return Empty properties            
2024-10-05 12:13:24.277 [DEBUG] [rnal.protocol.RemoteControllerLegacy] - 192.168.1.35: Open connection to host '192.168.1.35:55000'                                                           2024-10-05 12:13:24.281 [DEBUG] [rnal.protocol.RemoteControllerLegacy] - 192.168.1.35: Connection successfully opened...querying access
2024-10-05 12:13:24.324 [DEBUG] [rnal.protocol.RemoteControllerLegacy] - 192.168.1.35: Access granted                                   
2024-10-05 12:13:24.328 [DEBUG] [tv.internal.handler.SamsungTvHandler] - 192.168.1.35: Updated Configuration protocol:Legacy
2024-10-05 12:13:24.331 [DEBUG] [tv.internal.handler.SamsungTvHandler] - 192.168.1.35: Updated Configuration port:55000
2024-10-05 12:13:24.332 [TRACE] [tv.internal.handler.SamsungTvHandler] - 192.168.1.35: PowerState set to: on
2024-10-05 12:13:24.332 [DEBUG] [tv.internal.handler.SamsungTvHandler] - 192.168.1.35: Configuration: Legacy, port: 55000, token: , MAC: dc:71:44:52:cc:eb, subscription: false
2024-10-05 12:13:24.333 [DEBUG] [tv.internal.handler.SamsungTvHandler] - 192.168.1.35: Check and create missing services
2024-10-05 12:13:24.333 [DEBUG] [tv.internal.handler.SamsungTvHandler] - 192.168.1.35: Received value 'artMode':'OFF' for thing 'samsungtv:tv:eb938f0dc2'
2024-10-05 12:13:24.333 [DEBUG] [tv.internal.handler.SamsungTvHandler] - 192.168.1.35: Received value 'power':'OFF' for thing 'samsungtv:tv:eb938f0dc2'
2024-10-05 12:13:24.333 [TRACE] [tv.internal.handler.SamsungTvHandler] - 192.168.1.35: PowerState set to: off
2024-10-05 12:13:24.333 [DEBUG] [tv.internal.handler.SamsungTvHandler] - 192.168.1.35: Received value 'artImage':'NULL' for thing 'samsungtv:tv:eb938f0dc2'
2024-10-05 12:13:24.334 [DEBUG] [tv.internal.handler.SamsungTvHandler] - 192.168.1.35: Received value 'artLabel':'' for thing 'samsungtv:tv:eb938f0dc2'
2024-10-05 12:13:24.334 [DEBUG] [tv.internal.handler.SamsungTvHandler] - 192.168.1.35: Received value 'sourceApp':'' for thing 'samsungtv:tv:eb938f0dc2'

```

After that the thing stays offline forever. When turning off and on the TV, no messages are logged. When disabling and then re-enabling the thing, the following messages are logged, but no communication occurs between the binding and the TV (verified via tcpdump).

```
==> openhab.log <==
2024-10-05 12:23:54.955 [DEBUG] [tv.internal.handler.SamsungTvHandler] - 192.168.1.35: Configuration: Legacy, port: 55000, token: , MAC: dc:71:44:52:cc:eb, subscription: false
2024-10-05 12:23:54.956 [DEBUG] [tv.internal.handler.SamsungTvHandler] - 192.168.1.35: Check and create missing services
2024-10-05 12:23:54.956 [DEBUG] [tv.internal.handler.SamsungTvHandler] - 192.168.1.35: Received value 'artMode':'OFF' for thing 'samsungtv:tv:eb938f0dc2'
2024-10-05 12:23:54.956 [DEBUG] [tv.internal.handler.SamsungTvHandler] - 192.168.1.35: Received value 'power':'OFF' for thing 'samsungtv:tv:eb938f0dc2'
2024-10-05 12:23:54.956 [TRACE] [tv.internal.handler.SamsungTvHandler] - 192.168.1.35: PowerState set to: off
2024-10-05 12:23:54.957 [DEBUG] [tv.internal.handler.SamsungTvHandler] - 192.168.1.35: Received value 'artImage':'NULL' for thing 'samsungtv:tv:eb938f0dc2'
2024-10-05 12:23:54.957 [DEBUG] [tv.internal.handler.SamsungTvHandler] - 192.168.1.35: Received value 'artLabel':'' for thing 'samsungtv:tv:eb938f0dc2'
2024-10-05 12:23:54.957 [DEBUG] [tv.internal.handler.SamsungTvHandler] - 192.168.1.35: Received value 'sourceApp':'' for thing 'samsungtv:tv:eb938f0dc2'
2024-10-05 12:23:54.957 [DEBUG] [tv.internal.handler.SamsungTvHandler] - 192.168.1.35: TV is OFFLINE

==> events.log <==
2024-10-05 12:23:54.958 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'samsungtv:tv:eb938f0dc2' changed from UNKNOWN to OFFLINE

```

## Your Environment

* Version used: OpenHAB 4.2.2
* OpenHAB runs in a docker container with --net=host
* With netcat I can see the expected multicast messages inside the docker container
* With tcpdump I can see various messages from the TV, but after the initial detection never any communication originating from the binding.
* Remote Control is enabled on the TV for the correct IP address

Any help would be very much appreciated.

**Comment by lsiepel** [comment]
It seems the TV is not detected by UPnP or the services can;t be created. The method below corrosponds to your logging and ends with `putOffline `
https://github.com/openhab/openhab-addons/blob/8fcf18fc28a3a1e28bd54e7a8edf766b078acd49/bundles/org.openhab.binding.samsungtv/src/main/java/org/openhab/binding/samsungtv/internal/handler/SamsungTvHandler.java#L604-L629

I can try to fix it with more logging, but @NickWaterton knows a lot more about this binding so maybe he can comment.

**Comment by NickWaterton** [comment]
@martin5233 

The issue is that the binding is not seeing the UpNP multicast messages from your TV. This is likely because you are running Openhab in a docker container.

When the Thing is first added, the binding probes the TV to try to discover what type it is. This is an old TV (2012), so it only supports legacy access via port 5500. The binding correctly detects this and configures the binding accordingly. Older TV’s like this do not support remotely turning On, only Off.

In order to detect when the TV is on, the binding listens for the multicast messages from the TV - if they are not being passed into the docker container properly, the binding will not be able to detect that the TV is on - in which case the TV services will not start.

You could try enabling the “subscription” option in the binding configuration. This attempts to subscribe to the TV UpNP services. If the UpNP protocol is being filtered out by the docker container, this may not work though.

Ultimately, the problem is caused by the fact that you are running OH in a docker container. If you can set up an OH instance directly on an OS (as an experiment), the binding should work. This would at least confirm that the problem is related to docker.

**Comment by martin5233** [comment]
@lsiepel, @NickWaterton: Thanks a lot for your quick replies. I have the docker container running with --net=host and I also checked with netcat and tcpdump, that the UPNP messages are received in the docker container. Therefore I don't think, that the docker container is the culprit here. I had also tried to enable the "subscription", but this didn't change anything. 
Is there a logger for UPNP service, which I can raise the log level for?

**Comment by martin5233** [comment]
@NickWaterton: You are right, OpenHAB doesn't see any of these devices. I enabled the two loggers via karaf, but all I get every second is:
```
2024-10-05 16:35:33.108 [TRACE] [org.jupnp.registry.Registry         ] - Maintaining registry...
2024-10-05 16:35:33.108 [TRACE] [org.jupnp.registry.Registry         ] - Executing pending operations: 0
```

However what I still do not understand: Why can I see the multicast messages with netcat and tcpdump both started inside the same container, that OpenHAB is running in, but OpenHAB itself doesn't see them?

**Comment by NickWaterton** [comment]
@martin5233 

I don’t know why OpenHab doesn’t see the UpNP messages, but I have come across this problem with docker before. I don’t know if it’s an issue with the docker compose file options, or something else.

I suspect that it’s something like this issue:
https://forums.docker.com/t/multicast-on-docker/106856

**Comment by martin5233** [comment]
@NickWaterton: Thanks for the link to that multicast docker issue. However, if you read the issue closely, it says, that it works with "network_mode = host", which is what I use.
In any case I am in the process of setting up a VM with OpenHAB installed without Docker. I will report my findings...

**Comment by NickWaterton** [comment]
Ok, it actually says that it works with `network_mode = host` for *everything* in the docker compose file. This is why I think it’s a problem with the docker compose file.

I’m fairly sure the binding will work with the new VM, I have a 2012 Samsung TV as well, and that does work.

**Comment by martin5233** [comment]
@NickWaterton I have tried with a VM, but it basically behaves the same way as with docker. But I suspect, that the issue is slightly different there, because the VM is not on the same subnet as the TV. Inside that VM I don't receive any UPNP messages with netcat, which works inside the Docker container. I guess I will give up on this. The most important information for me is, when the TV is turned on and off, which I could also detect with the network binding. Thanks a lot for all your help!

**Comment by NickWaterton** [comment]
@martin5233 

It may be that the issue is not UpNP itself, but the multicast messages. It may be that the Linux kernel needs to have multicast routing enabled, and it’s disabled by default.

So UpNP messages, discovery etc may be sent (and you can see them), but *multicast* messages are not forwarded by the kernel, because the routing option is disabled.

I may take a closer look at this, as a couple of people have had the same problem, and while I’m sure it’s not a problem with the binding itself, it would be nice to know what the issue actually is.

**Comment by martin5233** [comment]
Hi @NickWaterton 

I just wanted to let you know, that I investigated the docker issue in conjunction with the SamsungTV binding a bit further. Here is what I found out:

- I can start OpenHAB in a docker container with host networking and the binding works as expected, if there are no other docker services running on the same machine. It's not even necessary to set all services defined in a docker-compose file to use host networking, doing this for OpenHAB is enough.
- When I start other docker services, which are completely unrelated to UPNP and OpenHAB, their IP addresses show up in OpenHAB's network configuration. But even when I choose the host address as the primary network interface, a previously detected TV is no longer recognized.

My suspicion is, that the additional network interfaces set up by Docker confuse OpenHAB, so it is listening on the wrong interface. With Wireshark I could see, that the initial discovery packets are still sent, the answers are received, but the reactions inside OpenHAB are missing. I tried to find out in the source code, where the actual listen call is made, but I'm not a Java guy and got lost somewhere in the UPNPIOService :-)

I hope that these observations might help to find the root cause of the problem.
